Dynamic reset time

zachmprince opened the pull request:
Closes #31034 

Application fixes:

- [ ] https://github.inl.gov/ncrc/sockeye/pull/873 (submodule update required)

zachmprince committed:
Support a list of number as a valid Times object, which gets create on-the-fly #31034

zachmprince committed:
Adding interface for grabbing a pointer to a times object #31034

zachmprince committed:
Making reset times in MultiApp dynamic with Times #31034

zachmprince committed:
Updating old newsletter mentioning sync_times_object #31034

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/3044009/), step Docs: sync website on 0029762 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/31046/site)

This comment will be updated on new commits.

zachmprince committed:
Modifying Times unit tests with changes #31034

zachmprince committed:
Fixing various test failures #31054

zachmprince committed:
Do not restart sub-apps when they are being reset

zachmprince committed:
Throwing error when default times are not sorted and adding error tests #31054

zachmprince committed:
Changing narrative in multiapp reset to make more sense #31034

GiudGiud commented on framework/src/multiapps/MultiApp.C:
```diff
@@ -662,41 +654,58 @@ MultiApp::preTransfer(Real /*dt*/, Real target_time)
+    // Not time to reset yet
+    if (prev_reset_time > _app.getStartTime())
+    {
+      // Check if we already reset this time
```
I like it!

zachmprince committed:
Allowing output sync_times to be unsorted #31034

GiudGiud commented:
@zachmprince can I take this over? This was a great idea

GiudGiud committed:
Move dynamic sequence capability to the have a parameter in user-facing times reporter class
And be an argument when constructing the backend times object

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3558405/), step Results summary on f653275 wanted to post the following:

## Framework test summary

Compared against 4b9b580 in job [civet.inl.gov/job/3550358](https://civet.inl.gov/job/3550358).

### Removed tests

| Test                                                          | Time (s)   | Memory (MB)   |
|---------------------------------------------------------------|------------|---------------|
| `time_steppers/timesequence_stepper.time_sequence/not_sorted` | 0.70       | 0.00          |
| `multiapps/reset.not_sorted_times`                            | 0.66       | 0.00          |

### Added tests

| Test                                           | Time (s)   | Memory (MB)   |
|------------------------------------------------|------------|---------------|
| `multiapps/reset.dynamic_reset`                | 1.40       | 0.00          |
| `outputs/intervals.sync_times_unsorted`        | 1.02       | 0.00          |
| `times.error_reporting/default_times_unsorted` | 0.71       | 0.00          |
| `times.error_reporting/not_times`              | 0.70       | 0.00          |
| `times.error_reporting/no_object`              | 0.70       | 0.00          |
## Modules test summary

Compared against 4b9b580 in job [civet.inl.gov/job/3550358](https://civet.inl.gov/job/3550358).

No change

GiudGiud committed:
Revert some previous test changes in THM

GiudGiud committed:
Use deprecation to avoid requiring app patches

GiudGiud committed:
Avoid adding the end_time to the time sequence
to prevent recovering with an additional time to hit

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3577014/), step Results summary on ca35de3 wanted to post the following:

## Framework test summary

Compared against 019760d in job [civet.inl.gov/job/3572357](https://civet.inl.gov/job/3572357).

### Removed tests

| Test                                                          | Time (s)   | Memory (MB)   |
|---------------------------------------------------------------|------------|---------------|
| `time_steppers/timesequence_stepper.time_sequence/not_sorted` | 0.70       | 0.00          |
| `multiapps/reset.not_sorted_times`                            | 0.66       | 0.00          |

### Added tests

| Test                                           | Time (s)   | Memory (MB)   |
|------------------------------------------------|------------|---------------|
| `multiapps/reset.dynamic_reset`                | 1.16       | 0.00          |
| `outputs/intervals.sync_times_unsorted`        | 1.00       | 0.00          |
| `times.error_reporting/default_times_unsorted` | 0.70       | 0.00          |
| `times.error_reporting/no_object`              | 0.68       | 0.00          |
| `times.error_reporting/not_times`              | 0.65       | 0.00          |
## Modules test summary

Compared against 019760d in job [civet.inl.gov/job/3572357](https://civet.inl.gov/job/3572357).

No change

GiudGiud committed:
Allow a different sequence of times when restarting: a restart is not a recover
we could have changed simulation objectives completely

Avoid inserting a duplicate start_time in the sequence vector

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3577260/), step Results summary on 3b514f1 wanted to post the following:

## Framework test summary

Compared against 019760d in job [civet.inl.gov/job/3572357](https://civet.inl.gov/job/3572357).

### Removed tests

| Test                                                          | Time (s)   | Memory (MB)   |
|---------------------------------------------------------------|------------|---------------|
| `time_steppers/timesequence_stepper.time_sequence/not_sorted` | 0.70       | 0.00          |
| `multiapps/reset.not_sorted_times`                            | 0.66       | 0.00          |

### Added tests

| Test                                           | Time (s)   | Memory (MB)   |
|------------------------------------------------|------------|---------------|
| `multiapps/reset.dynamic_reset`                | 1.17       | 0.00          |
| `outputs/intervals.sync_times_unsorted`        | 0.95       | 0.00          |
| `times.error_reporting/no_object`              | 0.69       | 0.00          |
| `times.error_reporting/default_times_unsorted` | 0.65       | 0.00          |
| `times.error_reporting/not_times`              | 0.61       | 0.00          |
## Modules test summary

Compared against 019760d in job [civet.inl.gov/job/3572357](https://civet.inl.gov/job/3572357).

No change
